### PR TITLE
[KAT-165] Optional Views and Topologies for PropertyGraph (also included Typed Property Graph Views)

### PR DESCRIPTION
**Pull request by amberhassaan**
This PR:
-- expands on optional topologies
-- wraps optional topologies into PropertyGraph Views
-- Creates a mechanism for caching topologies and other state to create Property Graph Views
-- Updates existing analytics applications to use PropertyGraphViews

Dependent PR: https://github.com/KatanaGraph/katana-enterprise/pull/1687

### CODE REVIEW

File: libgalois/include/katana/GraphTopology.h [hunk 1]
```diff
@@ -0,0 +1,1175 @@
...
+struct PropertyGraphViews {
+  using BiDirectional = PGViewBiDirectional;
+  using EdgesSortedByDestID = PGViewEdgesSortedByDestID;
+  using EdgeTypeAwareBiDir = PGViewEdgeTypeAwareBiDir;
+  using NodesSortedByDegreeEdgesSortedByDestID =
+      PGViewNodesSortedByDegreeEdgesSortedByDestID;
+};
+
+template <typename PGView>
+struct PGViewBuilder {};
```
Comment (arthurp) [inline libgalois/include/katana/GraphTopology.h:983]: I don't like the contraction `PG` for types in the `katana` namespace. Can we move these into `katana::internal` if they belong there or expand the contraction?

**Review by arthurp — APPROVED** [review]
LGTM. Other than how I don't want to publicly export the `PG` contraction.

File: libgalois/include/katana/GraphTopology.h (hunk 1 above)
Comment (amberhassaan) [reply libgalois/include/katana/GraphTopology.h:983]: Yes, we can.

File: libgalois/include/katana/GraphTopology.h [hunk 2]
```diff
@@ -0,0 +1,1175 @@
...
+namespace katana {
+
+class KATANA_EXPORT PropertyGraph;
+
+/// Types used by all topologies
+struct KATANA_EXPORT GraphTopologyTypes {
+  using Node = uint32_t;
+  using Edge = uint64_t;
+  using PropertyIndex = uint64_t;
+  using EntityType = uint8_t;
```
Comment (roshandathathri) [inline libgalois/include/katana/GraphTopology.h:28]: Shouldn't this be `using EntityType = EntityTypeID`?

File: libgalois/include/katana/GraphTopology.h [hunk 3]
```diff
@@ -0,0 +1,1175 @@
...
+
+namespace katana {
+
+class KATANA_EXPORT PropertyGraph;
+
+/// Types used by all topologies
+struct KATANA_EXPORT GraphTopologyTypes {
+  using Node = uint32_t;
+  using Edge = uint64_t;
+  using PropertyIndex = uint64_t;
```
Comment (roshandathathri) [inline libgalois/include/katana/GraphTopology.h:27]: Both node and edge property indices are 64-bit. Why? It's better to have `NodePropertyIndex` be uint32_t.

File: libgalois/include/katana/GraphTopology.h [hunk 4]
```diff
@@ -0,0 +1,1175 @@
...
+public:
+  GraphTopology() = default;
+  GraphTopology(GraphTopology&&) = default;
+  GraphTopology& operator=(GraphTopology&&) = default;
+
+  GraphTopology(const GraphTopology&) = delete;
+  GraphTopology& operator=(const GraphTopology&) = delete;
+
+  GraphTopology(
+      const Edge* adj_indices, size_t numNodes, const Node* dests,
```
Comment (roshandathathri) [inline libgalois/include/katana/GraphTopology.h]: `num_nodes`

File: libgalois/include/katana/GraphTopology.h [hunk 5]
```diff
@@ -0,0 +1,1175 @@
...
+      size_t numEdges) noexcept;
```
Comment (roshandathathri) [inline libgalois/include/katana/GraphTopology.h]: `num_edges`

File: libgalois/include/katana/GraphTopology.h [hunk 6]
```diff
@@ -0,0 +1,1175 @@
...
+  EdgeShuffleTopology(const EdgeShuffleTopology&) = delete;
+  EdgeShuffleTopology& operator=(const EdgeShuffleTopology&) = delete;
+
+  enum class TransposeKind : int { kNo = 0, kYes };
+
+  enum class EdgeSortKind : int {
+    kAny = 0,  // don't care. Sorted or Unsorted
+    kSortedByDestID,
+    kSortedByEdgeType,
+    kSortedByNodeType
```
Comment (roshandathathri) [inline libgalois/include/katana/GraphTopology.h:176]: Both edge and node type will have multiple elements with the same "order", so how ties are broken is important to be captured in the "Kind", right? I mean it should be something like `kSortedByEdgeTypeAndID`

File: libgalois/include/katana/GraphTopology.h [hunk 7]
```diff
@@ -0,0 +1,1175 @@
...
+  }
+
+  const PropertyGraph& property_graph() const noexcept { return *prop_graph_; }
+
+  PropertyIndex edge_property_index(const Edge& eid) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(eid < num_edges());
+    return edge_prop_indices_[eid];
+  }
+
+  void sortEdges(const EdgeSortKind& edge_sort_todo) noexcept {
```
Comment (roshandathathri) [inline libgalois/include/katana/GraphTopology.h]: `SortEdges`

File: libgalois/include/katana/GraphTopology.h [hunk 8]
```diff
@@ -0,0 +1,1175 @@
...
+
+    ret->sortEdges(edge_sort_todo);
+    return ret;
+  }
+
+  void SortEdgesByDestID() noexcept;
+
+  void SortEdgesByTypeThenDest() noexcept;
+
+  void SortEdgesByDestType(const PropIndexVec& node_prop_indices) noexcept;
```
Comment (roshandathathri) [inline libgalois/include/katana/GraphTopology.h]: These are public, so will they set the right state automatically?

File: libgalois/include/katana/GraphTopology.h [hunk 9]
```diff
@@ -0,0 +1,1175 @@
...
+        tpose_state_(tpose_todo),
+        edge_sort_state_(edge_sort_todo),
+        edge_prop_indices_(std::move(edge_prop_indices)) {
+    KATANA_LOG_DEBUG_ASSERT(edge_prop_indices_.size() == num_edges());
+  }
+
+private:
+  const PropertyGraph* prop_graph_;
+  bool is_valid_ = true;
+  TransposeKind tpose_state_ = TransposeKind::kNo;
```
Comment (roshandathathri) [inline libgalois/include/katana/GraphTopology.h:278]: nit: prefer full name `transpose_state_`

File: libgalois/include/katana/GraphTopology.h [hunk 10]
```diff
@@ -0,0 +1,1175 @@
...
+  EdgeSortKind edge_sort_state_ = EdgeSortKind::kAny;
+
+  // TODO(amber): In the future, we may need to keep a copy of edge_type_ids in
+  // addition to edge_prop_indices_. Today, we assume that we can use
+  // PropertyGraph.edge_type_set_id(edge_prop_indices_[edge_id]) to obtain
+  // edge_type_id. This may not be true when we group properties
```
Comment (roshandathathri) [inline libgalois/include/katana/GraphTopology.h:284]: edge_type_id will also be grouped in that case. If we group by type-id, then we don't need to maintain edge_type_id

File: libgalois/include/katana/GraphTopology.h [hunk 11]
```diff
@@ -0,0 +1,1175 @@
...
+
+  /// Wrapper to get the distinct edge types in the graph.
+  ///
+  /// @returns Range of the distinct edge types
+  EdgeTypeIDRange distinct_edge_type_ids() const noexcept {
+    return EdgeTypeIDRange{
+        index_to_type_map_.cbegin(), index_to_type_map_.cend()};
+  }
+
+  bool is_valid() const noexcept { return is_valid_; }
```
Comment (roshandathathri) [inline libgalois/include/katana/GraphTopology.h:488]: nit: `IsValid` is the norm in the rest of the code

File: libgalois/include/katana/GraphTopology.h [hunk 12]
```diff
@@ -0,0 +1,1175 @@
...
+  auto edges(const Node& N) const noexcept { return Base::edges(N); };
+
+  /// @param N node to get degree for
+  /// @param edge_type edge_type to get degree of
+  /// @returns Degree of node N
+  size_t degree(Node N, const EntityType& edge_type) const noexcept {
+    return edges(N, edge_type).size();
+  }
+
+  // C++ Derived classes hides Base class methods with the same name
```
Comment (roshandathathri) [inline libgalois/include/katana/GraphTopology.h:628]: I though this is only if there is a polymorphic overload in the derived class?

**Review by roshandathathri — COMMENTED** [review]
Partial comments

**Comment by amberhassaan** [comment]
@arthurp , @roshandathathri , @scober , @gurbinder533 : I've created a PR on top of this PR (https://github.com/amberhassaan/katana/pull/1) and sent out invitations to you guys (via github). Please review the changes in the  PR linked. I'll fix the build errors etc., but wanted to have you guys look at the changes

**Pull request merged**